src/components/common/RestEntity/README.md

```diff
@@ -0,0 +1,882 @@
+
+```javascript
+ React.useEffect(() => {
+    axiosAuth().get(`/programs/program/${props.programId}`)
+    .then(res => {
+      setData(res.data);
+    })
+    .catch(err => {
+      setError(err.message);
+    });
+  }, []);
+```
+
+That effect:
+
+- FETCH — makes a GET request at an endpoint `"/programs/program/${props.programId}"`
+
+- SUCCESS — sets data in `.then`
+
+- ERROR — snags an error message in `.catch`
+
+- LOADING — this is less visible, but the `async` nature of a `axios.get` implicitly includes "loading", as the Promise returned from the `.get()` will either SUCCEED or FAIL (error) -- but it's definitely going to RESOLVE. So BEFORE it resolves is when we're loading.
+
+The four parts of a fully implemented `RestEntity` align EXACTLY with that same approach.
+
+```javascript
+React.useEffect(() => {
+  axios.get(/* FETCH — "endpoint" */)
+  .then(/* SUCCESS — data obtained */)
+  .catch(/* FAIL — some error happened */)
+  .finally(/* RESOLVE */)
+}, []);
+
+<RestEntity href="endpoint"> {/*FETCH — .get("endpoint")*/}
+  <Singleton />  {/* SUCCESS — data obtained */}
+  <Error /> {/* FAIL — some error happened */}
+  <Loading /> {/* RESOLVE */}
+</RestEntity>
+```
+
+<br>
+
+> ##  Yup. Our `RestEntity` component is the JSX equivalent of an `axios` request.
+
+<br>
+
+Our `RestEntity` parent component takes in a `prop.href` to decide where the data it's supposed to fetch is located. What endpoint did we use in `ProgramComponent.js`?? This was located in the `useEffect` hook:
+
+```javascript
+axiosAuth().get(`/programs/program/${props.programId}`)
+```
+
+So... let's try this out:
+
+```javascript
+import React from "react";
+// import axios from "axios";
+
+import RestEntity from "./RestEntity";
+
+const ProgramComponent = props => {
+  // const [data, setData] = React.useState(undefined);
+  // const [error, setError] = React.useState(undefined);
+
+  // React.useEffect(() => {
+  //   axiosAuth().get(`/programs/program/${props.programId}`)
+  //   .then(res => {
+  //     setData(res.data);
+  //   })
+  //   .catch(err => {
+  //     setError(err);
+  //   });
+  // }, []);
+
+  return (
+    <RestEntity href={`/programs/program/${props.programId}`}>
+      {/*
+        okay... so what the hell did this do? 
+        Now the RestEntity has supposedly fetched my data.
+        But how do I access anything with it? 
+      */}
+      <RestEntity.Singleton />
+      <RestEntity.Error />
+      <RestEntity.Loading />
+    </RestEntity>
+  );
+};
+
+export default ProgramComponent;
+```
+
+Our `<RestEntity href=“/endpoint”>` parent component is going to take in that `href` prop and hit `GET /programs/program/${programId}` . Bang. Boom. In one line of code, you have effectively triggered a GET request to an endpoint. For now, that’s all you need to know about ***why*** that hits an endpoint: because you told it to.
```

Comment:
Overall you may want to divide these into more clear sections as the entire doc is a bit long and jumps between topics sometimes. Maybe you can define clear steps at the top, introduce those steps briefly, and then go through them in detail and in order for the rest of the doc. That way the reader has a clear point to refer back to at each stage.

Reply:
love it. and totally agree.